File: src/foam/swift/parse/json/ArrayParser.js
/**
 * @license
 * Copyright 2017 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

foam.CLASS({
  package: 'foam.swift.parse.json',
  name: 'ArrayParser',
  extends: 'foam.swift.parse.parser.ProxyParser',
  requires: [
    'foam.swift.parse.json.Whitespace',
    'foam.swift.parse.parser.Literal',
    'foam.swift.parse.parser.Repeat',
    'foam.swift.parse.parser.Seq0',
    'foam.swift.parse.parser.Seq1',
  ],
  axioms: [
    foam.pattern.Singleton.create()
  ],
  properties: [
  {
    name: 'delegate',
    swiftFactory: function() {/*
return Seq1_create(["index": 3, "parsers": [
  Whitespace_create(),
  Literal_create(["string": "["]),
  Whitespace_create(),
  Repeat_create([
    "delegate": __context__.create(foam_swift_parse_json_AnyParser.self)!,
    "delim": Seq0_create(["parsers": [
      Whitespace_create(),
      Literal_create(["string": ","]),
      Whitespace_create(),
    ]]),
  ]),
  Whitespace_create(),
  Literal_create(["string": "]"]),
]])
    */},
  },
  ],
});

File: src/foam/lib/json/ArrayParser.java
/**
 * @license
 * Copyright 2017 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

package foam.lib.json;

import foam.lib.parse.*;

public class ArrayParser
  extends ProxyParser
{
  private final static Parser instance__ = new ArrayParser();

  // Unusual definition of instance() to avoid circular references.
  public static Parser instance() { return instance__ == null ? new ProxyParser() { public Parser getDelegate() { return instance__; } } : instance__; }

  public ArrayParser() {
    super(new Seq1(3,
      Whitespace.instance(),
      Literal.create("["),
      Whitespace.instance(),
      new Repeat(
        AnyParser.instance(),
        new Seq0(Whitespace.instance(), Literal.create(","), Whitespace.instance())),
      Whitespace.instance(),
      Literal.create("]")));
  }
}
